### ISSUE TITLE
Grid moves and shifts relative to drawing contents on changing zoom

### ISSUE THREAD

**Issue by davidrea** [jgraph/drawio#212]
Steps to reproduce:

1. Create a new drawing, set grid to 5 pt.
1. Draw one or more shapes whose edges and vertices lay cleanly on grid lines.
1. Adjust the zoom to various levels; at some levels, the grid and object edges will no longer line up.
1. Move the objects; they no longer snap to grid lines or intersections, landing in whitespace.
1. If objects are moved at a zoom level where they do not align with the grid, their offsets relative to the grid persist at zoom levels where they previously aligned, making it impossible to reposition them back onto the grid.

This occurs in both web and desktop draw.io versions, and persists across closure/re-open of the same file. Previously-aligned items remain aligned, but groups can be pushed off-grid by moving them at the "wrong" zoom level. Objects' offset from the grid appears to be cyclic as the zoom level is changed over a large range.

I am running Draw.io 8.0.2 on Ubuntu Linux 16.04LTS with Chrome 63.0.3239.132.

**Comment by vblanshey** [comment]
It's been some time since this bug was reported but I have trouble due to the same problem. I am using the latest release 10.6.3  . While making a timing waveform diagram it is important to match edges of the waveforms to the grid points but once the zoom is used those points get moved from the grid. To keep with the grid points I cannot use zoom.

**Comment by mararad** [comment]
@vblanshey What OS are you using? At my place (draw.io online 11.1.2, Win10 and Chrome Version 74.0.3729.131 (Official Build) (64-bit) all looks ok.

**Comment by Ogi22po** [comment]
I have the same problem.
It would be useful if it was attached to the grid and not outside the grid.
The grid should remain in the same place during all adjustments and zooms.
It would be good to be inspired by AutoCad

(Draw.io v 12.3.8 - Google Chrome V 78.0.3904.108 (official) (64-bit) - Win10 64b

**Comment by davidjgraph** [comment]
What about with 12.4.2?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by TRSx80** [comment]
I have very similar problem.  I did not re-size though.  Steps to re-produce:

1. Create detailed drawing very carefully with lots of nodes all aligned to grid.
2. Save my work (yesterday).
3. Re-open (today).
4. Go to move something and realize all previous work is no longer aligned to grid.
5. OCD_TRIGGERED_ani.gif

I am using Desktop, version 12.6.5.

**Comment by MykhailoMe** [comment]
I have the same issue on new versions. May be it is due to I used old .drawio files?
Draw.io version 14.9.2

MacOs 10.15.7 Safari 13.1.3.
Tested also:
Windows 7 Opera 50, Firefox 68.12

If I may, it is not so easy to draw with a such bug...
This bug is known since 2018.
#212

Thank you for great tool!

My best wishes to Team and community!

**Comment by CelesteGS** [comment]
I also am having this issue.  Version is diagrams.net 20.2.8.  I am using Windows 10 Enterprise.

Agree that this bug makes it very difficult to use draw.io!

I love draw.io, thanks for all work on it :)

**Comment by davidjgraph** [comment]
We only deal with reports on the latest releases.

**Comment by CelesteGS** [comment]
Yes fair enough!  (I've been trying to get the latest version approved by my organisation, but don't know how likely that will be, so may only be able to use this version unfortunately)

**Comment by CelesteGS** [comment]
I've tried using the web version, 24.4.10, and the same issue is still occurring.

**Comment by davidjgraph** [comment]
can anyone show us a video?

**Comment by Archer6621** [comment]
@davidjgraph 

https://github.com/user-attachments/assets/cabde3b9-91fa-4c48-9039-0ebac66e55e5

Here you go. That's on the web version of drawio, on Firefox 133.0.3, Windows 10 Enterprise 22H2 build 19045.4651.

I did use some custom configuration, that is also shown in the video. I tested briefly on default values, and the problem still occurs, it is just significantly less visible. On these settings that I added, it is actually problematic and it makes the grid rather useless. 

From what I could tell, it seems that it is really more a visual issue than the actual grid shifting, as it doesn't properly snap to the shown grid.

EDIT: In addition, it seems that the shifting only happens on certain zoom levels.

**Comment by alderg** [comment]
What version of draw.io is this?

**Comment by Archer6621** [comment]
> @davidjgraph
> 2024-12-29.15-02-07.mp4
> 
> Here you go. That's on the web version of drawio, on Firefox 133.0.3, Windows 10 Enterprise 22H2 build 19045.4651.
> 
> I did use some custom configuration, that is also shown in the video. I tested briefly on default values, and the problem still occurs, it is just significantly less visible. On these settings that I added, it is actually problematic and it makes the grid rather useless.
> 
> From what I could tell, it seems that it is really more a visual issue than the actual grid shifting, as it doesn't properly snap to the shown grid.
> 
> EDIT: In addition, it seems that the shifting only happens on certain zoom levels.

I found that if you put the grid size (in the page setup) to `2.5` (I think it will round it to `2` on its own), without any further configuration tweaks, massive shifting around of the grid will occur when zooming in/out.

**Comment by Archer6621** [comment]
> What version of draw.io is this?

The web version that is hosted here: https://app.diagrams.net

Currently: `v26.0.2`

**Comment by jalinei** [comment]
Any update on this ? It is one of the most annoying problem when using draw.io...

**Comment by alderg** [comment]
We're unable to reproduce the problem. Can someone please give a sample diagram and a zoom level where this happens?

**Comment by dmitry-smith-snlpm** [comment]
<img width="990" height="649" alt="Image" src="https://github.com/user-attachments/assets/f5fc3607-4a27-4a6f-8ca6-f018f1805457" />
same thing, mismatch in grid and shapes contour

40%, 60%, 70%, 85%, 210%

v28.1.2 standalone, win 11

**Comment by dmitry-smith-snlpm** [comment]
<img width="300" height="284" alt="Image" src="https://github.com/user-attachments/assets/1af66092-b94c-4fc7-ba24-f966062acf60" />

[simple diagram.drawio](https://github.com/user-attachments/files/22902231/simple.diagram.drawio)

diagram in attachment

**Comment by dmitry-smith-snlpm** [comment]
> We're unable to reproduce the problem. Can someone please give a sample diagram and a zoom level where this happens?

Hi! Are these screenshots enough to repeat the situation and identify the cause of the problem?

**Comment by davidjgraph** [comment]
@dmitry-smith-snlpm thanks for the sample file and zoom levels — that made this pin-downable.

What the screenshots show isn't the grid actually shifting: the canvas grid is painted as a repeating SVG tile, and the major grid line was stroked along the tile's top/left boundary, where the pattern clips the outer half of the stroke. Every painted major line was therefore only half a pixel wide, visually offset ~0.25px from the true grid position (minor lines were unaffected). At 100% display scaling that can read as shape outlines sitting up to a pixel off the grid at various zoom levels, which matches your screenshots. Shape geometry itself is never off-grid — snapping happens in model coordinates, unaffected by zoom.

Fixed in dev by e0c5ef92d700ea9a6370b63448dc0447e0941914: the major line is now stroked on both tile edges, so adjacent repeats reconstruct a full, centered stroke. Verified against your file at 40/60/70/85/210% (and 121/144%) — painted major lines now measure dead-center on the model grid. This will be in the next release.

The older reports here of the grid drifting progressively across the page at certain zoom levels (e.g. 121%/144%) were a browser limitation when tiling fractional-size backgrounds; current Chrome/Electron handles this correctly (verified ~200 tile repeats from the page origin with no accumulated drift). If anyone still sees multi-pixel drift on a current browser once the next release is out, please open a fresh report with your browser version and OS display scaling.
